### PR TITLE
recognize async main methods

### PR DESCRIPTION
**Pull request by TyOverby**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 1]
```diff
@@ -1104,7 +1104,6 @@ internal enum ErrorCode
         ERR_VarargsAsync = 4006,
         ERR_ByRefTypeAndAwait = 4007,
         ERR_BadAwaitArgVoidCall = 4008,
-        ERR_MainCantBeAsync = 4009,
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: i think convention is to comment out, so no one uses this number again.

Comment (TyOverby) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: PR FOR PERSONAL USE ONLY

Comment (TyOverby) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: Just kidding.  Thank you!

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs [hunk 2]
```diff
@@ -1307,6 +1307,27 @@ internal static bool IsGenericTaskType(this TypeSymbol type, CSharpCompilation c
...
         }
 
+        internal static bool IsAsyncMainReturnType(this TypeSymbol type, CSharpCompilation compilation)
+        {
+            var namedType = type as NamedTypeSymbol;
+            if (type.IsNonGenericTaskType(compilation))
+            {
+                return true;
+            }
+            else if ((object)namedType == null || namedType.Arity != 1)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: arity check seems superfluous due to later checks which will validate that.

**Comment by CyrusNajmabadi** [comment]
Oops :) didn't see that.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs (hunk 2 above)
Comment (TyOverby) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: @CyrusNajmabadi: This is checking the arity of the type.  That is, `Foo<T, E>` has arity 2.  

However, I've restructured this method to be a bit more clear.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs [hunk 3]
```diff
@@ -1307,6 +1307,28 @@ internal static bool IsGenericTaskType(this TypeSymbol type, CSharpCompilation c
...
+            {
+                return false;
+            }
+            else if (namedType.IsNonGenericTaskType(compilation))
+            {
+                return true;
+            }
+            else if ((object)namedType.ConstructedFrom == compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task_T))
+            {
+                return namedType.TypeParameters.Length == 1 &&
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: this check is unnecessary.  If your ```ConstructedFrom == Task<>```, then your TypeParameter.Length is 1 by invariance.

Comment (TyOverby) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: Ah yep.  Good point

**Comment by TyOverby** [comment]
Closing in favor of https://github.com/dotnet/roslyn/pull/18157

**Pull request closed without merging**